### ISSUE TITLE
PF4: Add unique ID to FormGroup children/label

### ISSUE THREAD

**Issue by dlabrecq** [patternfly/patternfly-react#799]
The koku-ui implemented its own FormGroup component, prior to PF4, which assigns a unique ID to its children and label. I would like to suggest that this feature is also added to the PF4 FormGroup component, if an ID has not been provided.

Koku-ui FormGroup
https://github.com/project-koku/koku-ui/blob/master/src/components/formGroup/formGroup.tsx#L12

Koku-ui RandomId
https://github.com/project-koku/koku-ui/blob/master/src/components/randomId/randomId.tsx

**Comment by martinpovolny** [comment]
Why don't you just set the propTypes on the child element? It's the child's responsibility to have the ID.

**Comment by dlabrecq** [comment]
Of course I can set the ID manually. I just thought this would be a helpful feature, since FormGroup also outputs a label requiring the same ID.

**Comment by dmiller9911** [comment]
I agree with @dlabrecq here, although I might be biased since I wrote that random ID helper. Since we are trying to be accessible by default with PF we should make accessibility come easily to consumers. If all we do is warn with PropTypes it is possible for devs to ignore them. Throwing is even worse since devs will likely just drop it entirely. Additionally, move devs have no need for id's other than to make them accessible. So if we would allow fromGroup to do this automatically it is likely that we can promote more accessible apps without devs having to do anything. This will apply to modals, drop downs, and more as well since ID's will be required for those.

**Comment by martinpovolny** [comment]
1) This is violating the single responsibility principle by adding an extra responsibility to the FormGroup.
2) There already is a mechanism to declare what props are mandatory. That is the PropTypes.
3) It's a generaly bad practice for a component to inspect properties of it's child components.
4) With this idea the PF is trying to be smarter than the user of the library, the developer. Sounds familiar, do you remember Mr. Clippy? It's going to bite back.

![clippy](https://upload.wikimedia.org/wikipedia/en/d/db/Clippy-letter.PNG)

I suggest trying to figure how to promote good usability w/o violating software design principles.

BTW: https://github.com/patternfly/patternfly-react/pull/823 Seems it is biting back already.

**Comment by himdel** [comment]
1. It seems trivial to implement your own FormGroupRandom which does that.
2. Not every patternfly form will use it/need it (in fact the idea is broken whenever you have more than one input in a form group (like, date & time))
3. There is already too much magic around certain PF parts (though usually it comes from react-bootstrap, not so sure about pf4 yet)
4. Accessibility is **not** as easy as adding random ids

So... Why should this be in patternfly instead of your code?

**Comment by Hyperkid123** [comment]
> Of course I can set the ID manually. I just thought this would be a helpful a11y feature, since FormGroup also outputs a label requiring the same ID.

I don't really understand how randomly generated ID will help with a11y. I thought that a11y should bring additional information to users. I can't imagine that any random string of characters or one generated by tag type or w/e would help anybody. Maybe you can clarify this.

I am all form a11y, but it should be primarily decided by the library consumer. The library should provide tools that will help implementing it. And if there are components that just simply must have them like form inputs or labels (not some style wrappers), just enforce it via props of THE component, but not some parent.

**Comment by martinpovolny** [comment]
> just enforce it via props of THE component, but not some parent.

^^^^
this

**Comment by martinpovolny** [comment]
> I don't really understand how randomly generated ID will help with a11y. I thought that a11y should bring additional information to users. I can't imagine that any random string of characters or one generated by tag type or w/e would help anybody. Maybe you can clarify this.

That's a really interesting question. I second that. Can @dlabrecq and/or @dmiller9911, please, explain that? Maybe we just lack some important knowledge in this topic what would allow us to see this issue in a different light.

To me this seems like a way to formally pass some a11y testing. I can (with an effort) imagine that a random ID plus a meaningful label can help a screenreader. But if fixing the problem was as easy as adding random generated IDs and labels then the screen reader software could do it. I am really interested in an explanation of this.

**Comment by dmiller9911** [comment]
The ids do help. They make it accessible by default. PropTypes are decent for informing the user, but they don't enforce anything. Making inputs be accessible should come easy. While passing an id is "easy" it can become unneeded if the component lib does this for the consumer. This is not much harder than adding an id on instantiation so I am not sure where the push back to complexity is here. This is also a pattern most modern and forward-thinking component libs are using. For example

Mineral-UI: https://github.com/mineral-ui/mineral-ui/blob/master/src/library/Form/FormField.js#L135
Mineral UI actually has a similar focus on accessibility that PatternFly does.

ReachUI - https://github.com/reach/reach-ui/blob/master/packages/menu-button/src/index.js#L68
Reach UI is built and maintained by two thought leaders in the React Community, Ryan Florence and Kent Dodds. The only mission of this lib is accessibility. 

Since I probably should address the quotes my answeres to them are below.

> I don't really understand how randomly generated ID will help with a11y. I thought that a11y should bring additional information to users. I can't imagine that any random string of characters or one generated by tag type or w/e would help anybody. Maybe you can clarify this.

Ids are needed for labels and inputs, as well as other various parts of the form group. They must all match, but they don't actually need to be useful. The screen reader uses them to reference the label, errors, etc for the input. Since we are not wrapping the input in the label element we have to pass an `htmlFor` (`for` in HTML). the value of this id does not matter it must just be unique on the page. The downside to things currently is that is imperative that the ids match. Right now we have no way of validating or forcing this
```jsx
<FormGroup label="Label" fieldId="fieldID">
  <TextInput id="fieldId" />
</FormGroup>
```
See any problems with that? the Id's don't match and you got no warnings. If the FormGroup component manages ids everything will work. I get that we can say well that is up to the dev to not mess up. Sure that is a good argument for most things, but PatternFly's mission is to be accessible by default. We can accomplish this easily with the id.

> just enforce it via props of THE component, but not some parent.

Same as above. Id's need to match not just be present. We can't enforce that successfully. 

> This is violating the single responsibility principle by adding an extra responsibility to the FormGroup.

FormGroup already renders the label, the input, the error. Single responsibility should not get in the way here. I do agree that we should have components for each so that they are single responsibility, but this is already passed that. Adding an id to me does not add to the complexity here. As well this is a component built to a specific design system. There should only be a single child that is an input......if not then it doesn't follow the design guidelines for a `FormGroup`

> Not every patternfly form will use it/need it (in fact the idea is broken whenever you have more than one input in a form group (like date & time))

Yes, they will. Accessibility should be a requirement not an after thought. Date time will be handled differently anyways since it is much more complex than a FormGroup Component. A form group as defined by the design system does not have multiple inputs so again if someone does that it is already broken.

> It's a generally bad practice for a component to inspect properties of its child components.

Can agree with this one. Cloning an element with props actually happens way more than you think. React Router for instance: https://github.com/ReactTraining/react-router/blob/fe8d692f25bfa75189bcbf011ba534bc89090ea5/packages/react-router/modules/Switch.js#L40 That is really just one example, but it shows that some of the most used libs utilize this. That being said we wouldn't technically need to do this at a clone level. We could easily use react's context API to manage ids and have our input components consume that context.

> With this idea the PF is trying to be smarter than the user of the library, the developer. Sounds familiar, do you remember Mr. Clippy? It's going to bite back.

I do remember Clippy. To me warning via prop-types instead of just doing it by default is closer to that analog. Since he would notice you were trying something and then "warn" or notify the user about the fact they are failing. We are not trying to be "smarter" than the dev, but accessibility is not easy and is often an after thought for devs. We have the oppurtunity to make it accssibile by default not accessible by docs.

TL;DR: We should bake as much accessibility in by default instead of requiring the dev to do this. As long as well allow the dev to overwrite things we should be ok. This would likely be a ~10 line addition to make all inputs accessible by default for devs.

@jgiardino @seanforyou23  What are your thoughts on this. I assume we should be trying to avoid pushing accessibility to consumers, but I might be on the wrong page here.

**Comment by rvsia** [comment]
@dmiller9911

> There should only be a single child that is an input......if not then it doesn't follow the design guidelines for a FormGroup

> A form group as defined by the design system does not have multiple inputs so again if someone does that it is already broken.

I don't think this is true. FormGroup is designed to have more inputs, see [here](https://pf-next.com/demos/BasicForms/examples/): in the first example, there are three radios inputs in one inlined formGroup.

**Comment by himdel** [comment]
Yes, @dmiller9911 sounds like you're expecting FormGroup to address only a subset of the actual cases.

If the idea to use `<div className="form-group">` (whatever the pf4 class equivalent is) instead of `<FormGroup>` for all those other cases, please think about this again.

Again, if this makes sense for a subset of form group uses, this should be a separate component, not a part of the generic FormGroup.

**Comment by jschuler** [comment]
I also think that we should not make the id required, but optional, and in the cases where the user does not provide the optional id we should generate them and use them to bind the label and input together. In the case of the multiple radio button example, they also come from a component that should do the same (generating id if optional id is not passed in as prop).
If the user passes in id we should use that instead of course, but why not make it easier to use these components?

**Comment by martinpovolny** [comment]
Reading the discussion it seems to me that we are talking about 2 different components here.

One is a wrapper around something that forms a single logical form element. That seems to be what  @dmiller9911  is talking about and what the examples that he provided seem to be doing.

The 2nd component is the FormGroup that @rvsia actually developed for PF4. It's an element that visually groups a set of form elements in a more complex form. Such component exists in PF3 and we need such component for the data-driven forms component that mainly @Hyperkid123 is working on (and that 2 teams on insights are already using).

Seems to me that these probably need to be 2 different things and that mixing them 2 together has caused the confusion and disagreement around this issue and here: https://github.com/patternfly/patternfly-react/pull/823 , https://github.com/patternfly/patternfly-react/pull/667

@jschuler, @dmiller9911 : It what I wrote is the case then we should probably name the 2 differently and the disagreement would be resolved.

**Comment by dgutride** [comment]
For now, I'd recommend we allow the user to pass in the ID, when we have time to do a full investigation based on all of the comments here, I think we should revisit and provide a path forward.
